src/metaxy/metadata_store/delta.py

```diff
@@ -0,0 +1,484 @@
+"""Delta Lake metadata store implemented with delta-rs."""
+
+from __future__ import annotations
+
+import warnings
+from collections.abc import Iterator, Sequence
+from contextlib import contextmanager
+from functools import cached_property
+from pathlib import Path
+from typing import Any, Literal
+from urllib.parse import urlsplit
+
+import narwhals as nw
+import polars as pl
+
+from metaxy.metadata_store.base import MetadataStore
+from metaxy.metadata_store.exceptions import TableNotFoundError
+from metaxy.models.feature import BaseFeature
+from metaxy.models.plan import FeaturePlan
+from metaxy.models.types import FeatureKey
+from metaxy.provenance import ProvenanceTracker
+from metaxy.provenance.types import HashAlgorithm
+
+
+class DeltaMetadataStore(MetadataStore):
+    """
+    Delta Lake metadata store backed by [delta-rs](https://github.com/delta-io/delta-rs).
+
+    Stores each feature's metadata in a dedicated Delta table located under ``root_path``.
+    Uses Polars/Narwhals components for metadata operations, supports configurable storage
+    layouts (flat vs. nested directories), and forwards ``delta_write_options`` to every
+    delta-rs write call.
+
+    **Lazy Frame Handling:**
+
+    - **Eager DataFrames**: Written directly using Polars' native write_delta (efficient)
+
+    - **Lazy Frames with streaming_chunk_size=None** (default): Collected once then written
+
+    - **Lazy Frames with streaming_chunk_size=N**: Streamed in batches of size N without full collection,
+      enabling processing of datasets larger than RAM (at the cost of slower performance and more Delta versions)
+
+    Example:
+        ```py
+        # Standard usage (collects lazy frames before writing)
+        store = DeltaMetadataStore("/data/metaxy/metadata")
+
+        # Streaming mode for large datasets
+        store = DeltaMetadataStore(
+            "/data/metaxy/metadata",
+            streaming_chunk_size=10000,  # Stream in 10k row batches
+            storage_options={"AWS_REGION": "us-west-2"},
+        )
+
+        with store:
+            with store.allow_cross_project_writes():
```

Comment:
This allow_cross_project_writes thing doesn't belong here

Reply:
removed_